Mathlib/FieldTheory/RatFunc/Hilbert.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2024 Fangming Li. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fangming Li, Jujian Zhang
+-/
+import Mathlib.Algebra.Polynomial.AlgebraMap
+import Mathlib.Algebra.Polynomial.Eval.SMul
+import Mathlib.RingTheory.Polynomial.Pochhammer
+import Mathlib.RingTheory.PowerSeries.WellKnown
+import Mathlib.Tactic.FieldSimp
+
+/-!
+# Hilbert polynomials
+
+In this file, we formalise the following statement: if `F` is a field with characteristic `0`, then
+given any `p : F[X]` and `d : ℕ`, there exists some `h : F[X]` such that for any large enough
+`n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`.
+This `h` is unique and is denoted as `RatFunc.Polynomial.hilbert p d`.
+
+For example, given `d : ℕ`, the power series expansion of `1/(1-X)ᵈ⁺¹` in `F[X]` is
+`Σₙ ((d + n).choose d)Xⁿ`, which equals `Σₙ ((n + 1)···(n + d)/d!)Xⁿ` and hence
+`RatFunc.Polynomial.hilbert (1 : F[X]) d` is the polynomial `(n + 1)···(n + d)/d!`. Note that if
+`d! = 0` in `F`, then the polynomial `(n + 1)···(n + d)/d!` no longer works, so we do not want the
+characteristic of `F` to be divisible by `d!`. As `RatFunc.Polynomial.hilbert` may take any
+`p : F[X]` and `d : ℕ` as its inputs, it is necessary for us to assume that `CharZero F`.
+
+## Main definitions
+
+* `RatFunc.Polynomial.hilbert p d`. Given a field `F`, a polynomial `p : F[X]` and a natural number
+  `d`, if `F` is of characteristic `0`, then `RatFunc.Polynomial.hilbert p d : F[X]` is the
+  polynomial whose value at `n` equals the coefficient of `Xⁿ` in the power series expansion of
+  `p/(1 - X)ᵈ`.
+
+## TODO
+
+* Hilbert polynomials of finitely generated graded modules over Noetherian rings.
+-/
+
+open Nat Polynomial PowerSeries
+
+variable (F : Type*) [Field F]
+
+namespace RatFunc
```

Comment:
I suggest removing this from the namespace, so that Hilbert polynomials will be `Polynomial.hilbert`.
But I'm happy to hear counterarguments!

Reply:
@kbuzzard and I suggest that the definition shouldn't be `Polynomial.hilbert`. This is because the thing I am constructing is not the Hilbert polynomial of some graded module. Rather, it is more like "the Hilbert polynomial with respect to some polynomial", but that does not exist in literature.

Reply:
What has a Hilbert polynomial? A graded algebra, a graded module over that graded algebra, a projective scheme, a coherent sheaf on that projective scheme. I thought it was a bit arrogant to "claim" `Polynomial.hilbert` for this arguably auxiliary notion, which was why I suggested it went in `RatFunc` because it's all about writing down a polynomial whose value at $n$ is the coefficient of $X^n$ in a certain rational function $P(X)/(1-X)^d$.

Reply:
I would suggest `RatFunc.hilbertPoly` instead, because `RatFunc.Polynomial.hilbert` is quite long.

Reply:
> I would suggest `RatFunc.hilbertPoly` instead, because `RatFunc.Polynomial.hilbert` is quite long.

Yeah I feel that this is a great idea!

Reply:
I would be happy with `Polynomial.hilbertPoly` or `RatFunc.hilbertPoly`. I was initially imagining `Scheme.Polynomial.hilbert` and was going to suggest `Polynomial.Polynomial.hilbert` but this sets off a linter! I also was a bit unsure about what would happen after `open Polynomial` :-)

Reply:
I guess `Polynomial.hilbertPoly` would be something we all agree on. So may I just take `Polynomial.hilbertPoly` to be the final name?

Reply:
bors d+